core/src/main/java/org/apache/accumulo/core/fate/zookeeper/ServiceLockSupport.java

```diff
@@ -0,0 +1,152 @@
+   * service instance fails and loses its lock in ZK.
+   */
+  public static class HAServiceLockWatcher implements AccumuloLockWatcher {
+
+    private static final Logger LOG = LoggerFactory.getLogger(HAServiceLockWatcher.class);
+
+    private final String serviceName;
+    private volatile boolean acquiredLock = false;
+    private volatile boolean failedToAcquireLock = false;
+
+    public HAServiceLockWatcher(String serviceName) {
+      this.serviceName = serviceName;
+    }
+
+    @Override
+    public void lostLock(LockLossReason reason) {
+      Halt.halt(serviceName + " lock in zookeeper lost (reason = " + reason + "), exiting!", -1);
+    }
+
+    @Override
+    public void unableToMonitorLockNode(final Exception e) {
+      // ACCUMULO-3651 Changed level to error and added FATAL to message for slf4j compatibility
+      Halt.halt(-1,
+          () -> LOG.error("FATAL: No longer able to monitor {} lock node", serviceName, e));
+
+    }
+
+    @Override
+    public synchronized void acquiredLock() {
+      LOG.debug("Acquired {} lock", serviceName);
+
+      if (acquiredLock || failedToAcquireLock) {
+        Halt.halt("Zoolock in unexpected state AL " + acquiredLock + " " + failedToAcquireLock, -1);
+      }
+
+      acquiredLock = true;
+      notifyAll();
+    }
+
+    @Override
+    public synchronized void failedToAcquireLock(Exception e) {
+      LOG.warn("Failed to get {} lock", serviceName, e);
+
+      if (e instanceof NoAuthException) {
+        String msg =
+            "Failed to acquire " + serviceName + " lock due to incorrect ZooKeeper authentication.";
+        LOG.error("{} Ensure instance.secret is consistent across Accumulo configuration", msg, e);
+        Halt.halt(msg, -1);
+      }
+
+      if (acquiredLock) {
+        Halt.halt("Zoolock in unexpected state acquiredLock true with FAL " + failedToAcquireLock,
+            -1);
+      }
+
+      failedToAcquireLock = true;
+      notifyAll();
+    }
+
+    public synchronized void waitForChange() {
+      while (!acquiredLock && !failedToAcquireLock) {
+        try {
+          LOG.info("{} lock held by someone else, waiting for a change in state", serviceName);
+          wait();
+        } catch (InterruptedException e) {
+          // empty
+        }
+      }
+    }
+
+    public boolean isLockAcquired() {
+      return acquiredLock;
+    }
+
+    public boolean isFailedToAcquireLock() {
+      return failedToAcquireLock;
+    }
+
+  }
+
+  /**
+   * Lock Watcher used by non-HA services
+   */
+  public static class ServiceLockWatcher implements LockWatcher {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ServiceLockWatcher.class);
+
+    private final String serviceName;
+    private final Supplier<Boolean> shuttingDown;
+    private final Consumer<String> lostLockAction;
+
+    public ServiceLockWatcher(String serviceName, Supplier<Boolean> shuttingDown,
```

Comment:
On the merge to `main` I would probably change `serviceName` from a `String` to `ServerId.ServerType`.